design/resource-validation.md

```diff
@@ -0,0 +1,217 @@
+  updated from a valid to an invalid state, NKG will also ignore any previous valid state. For example, it will remove
+  the generation configuration for an HTTPRoute resource.
+* NKG will report the validation error as a
+  Warning [Event](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/)
+  for that resource. The Event message will describe the error and explain that the resource was ignored. We chose to
+  report an Event instead of updating the status, because to update the status, NKG first needs to look inside the
+  resource to determine whether it belongs to it or not. However, since the webhook validation applies to all parts of
+  the spec of resource, it means NKG has to look inside the invalid resource and parse potentially invalid parts. To
+  avoid that, NKG will report an Event.
+
+### NKG-specific validation
+
+After re-running the webhook validation, NKG will run NKG-specific validation, written in go.
+
+NKG-specific validation will:
+
+1. Ensure field values are considered valid by NGINX (cannot make NGINX fail to reload).
+2. Ensure valid field values do not include any malicious configuration.
+3. Report an error if an unsupported field is present in a resource (as the Gateway API spec prescribes).
+
+NKG-specific validation will not include:
+
+- *All* validation done by CRDs. NKG will only repeat the validation that addresses (1) and (2) in the list above with
+  extra rules required by NGINX but missing in the CRDs. For example, NKG will not ensure the limits of field values.
+- The validation done by the webhook (because it is done in the previous step).
+
+If a resource is invalid, NKG will report the error in its status.
+
+### Summary of Validation
+
+The table below summarizes the validation methods NKG will use. Any Gateway API resource will be validated by the
+following methods in order of their appearance in the table.
+
+| Name                         | Type    | Component             | Scope                   | Feedback loop for errors                                                         | Can be bypassed?                                                                     |
+|------------------------------|---------|-----------------------|-------------------------|----------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|
+| CRD validation               | OpenAPI | Kubernetes API server | Structure, field values | Kubernetes API server returns any errors a response for an API call.             | Yes, if the CRDs are modified.                                                       |
+| Webhook validation           | Go code | Gateway API webhook   | Field values            | Kubernetes API server returns any errors a response for an API call.             | Yes, if the webhook is not installed, misconfigured, or running a different version. |
```

Comment:
Any thoughts on extending the webhook validation to support NKG-specific validation? That way we can report errors back to user on CRUD events.

Reply:
This would be problematic because of the assumption that multiple Gateway implementations can simultaneously run in the cluster. Because a resource can be invalid from NKG perspective but still valid from another implementation perspective, running an NKG-specific webhook will mess up other Gateway implementations,

In this design, we assume that the cluster admin installs the Webhook provided by the Gateway API project.

Reply:
Makes sense for Gateway API resources. What about future NKG CRDs?

Reply:
Because we will have full control of those CRDs, I expect that CRD-based validation will cover most of the cases, so having our own Webhook won't worth it. However, if that expectation turns out incorrect, we can revisit and introduce our own Webhook.